internal/common/resource.go

```diff
@@ -132,6 +138,80 @@ func CreateOrUpdate(request *Request) ReconcileBuilder {
 	}
 }
 
+func Cleanup(request *Request, resource client.Object) (CleanupResult, error) {
+	found := newEmptyResource(resource)
+	err := request.Client.Get(request.Context, client.ObjectKeyFromObject(resource), found)
+	if errors.IsNotFound(err) {
+		return CleanupResult{
+			Resource: resource,
+			Deleted:  true,
+		}, nil
+	}
+	if err != nil {
+		return CleanupResult{}, err
+	}
+
+	// Add owner ref and annotations to compare with found object
+	resource.SetOwnerReferences(nil)
+	if resource.GetNamespace() == request.Instance.GetNamespace() {
+		err = controllerutil.SetControllerReference(request.Instance, resource, request.Client.Scheme())
+		if err != nil {
+			return CleanupResult{}, err
+		}
+	}
+	err = libhandler.SetOwnerAnnotations(request.Instance, resource)
+	if err != nil {
+		return CleanupResult{}, err
+	}
+
+	isOwned := false
+	for _, reference := range found.GetOwnerReferences() {
+		if reflect.DeepEqual(reference, resource.GetOwnerReferences()[0]) {
+			isOwned = true
+		}
+	}
+	if found.GetAnnotations() != nil {
```

Comment:
is possible to add here another check `&& !isOwned` ?

Reply:
Sure. This is one of the cases, when I cannot decide between readability and performance. Both improvements are very small.